src/nvim/os/input.c

```diff
@@ -323,13 +325,25 @@ static unsigned int handle_mouse_event(char **ptr, uint8_t *buf,
   return bufsize;
 }
 
+bool input_blocking(void)
+{
+  return blocking;
+}
+
 static bool input_poll(int ms)
 {
   if (do_profiling == PROF_YES && ms) {
     prof_inchar_enter();
   }
 
+  if ((ms == - 1 || ms > 0)
+      && !(events_enabled || input_ready() || input_eof)
```

Comment:
I was afraid of that. The only other option I can think of is to manage the `nvim_get_mode` requests in a child queue, and then process only those calls here. E.g.

    channel_process_safe_events()

or something like that. And that would iterate through `nvim_get_mode` events. 

I thought maybe `main_loop.fast_events` could be used for this purpose somehow, but didn't have success.